reconciler/healthchecks/healthchecks.go

```diff
@@ -107,122 +121,136 @@ func validateCheck(check *types.RealServer_HealthCheck) error {
 	return nil
 }
 
-func (c *checker) SetHealthCheck(serviceID string, key *types.RealServer_Key, check *types.RealServer_HealthCheck) error {
+func (c *checker) SetHealthCheck(serviceID string, key *types.RealServer_Key, check *types.RealServer_HealthCheck,
+	fn TransitionFunc) error {
+
 	if err := validateCheck(check); err != nil {
 		return fmt.Errorf("invalid healthcheck %v: %v", check, err)
 	}
 
+	if fn == nil {
+		fn = func(_ HealthState) {}
+	}
+
 	c.Lock()
 	defer c.Unlock()
 
-	infoKey := infoKey{id: serviceID, key: *key}
-	if info, ok := c.infos[infoKey]; ok {
-		// don't update if it's the same
-		if proto.Equal(info.healthCheck, check) {
-			return nil
-		}
-		c.updateHealthCheck(info, check)
-		return nil
+	status := &healthCheckStatus{
+		state: ServerDown,
+		count: 0,
 	}
 
-	info := &checkInfo{
-		serverIP:    key.Ip,
-		healthCheck: check,
-		status: &checkStatus{
-			state: serverDown,
-			count: 0,
-		},
-	}
-	c.infos[infoKey] = info
+	stateKey := healthCheckStateKey{id: serviceID, key: *key}
+	if origState, ok := c.healthCheckStates[stateKey]; ok {
 
-	info.startBackgroundHealthCheck()
-	return nil
-}
+		origState.Lock()
+		origState.transitionFn = fn
+		origState.Unlock()
 
-func (c *checker) updateHealthCheck(info *checkInfo, check *types.RealServer_HealthCheck) {
-	info.healthCheck = check
+		// Don't replace health check if it's the same.
+		if proto.Equal(origState.healthCheck, check) {
+			return nil
+		}
 
-	// stop any existing health check goroutine
-	if info.stopCh != nil {
-		close(info.stopCh)
-		info.stopCh = nil
+		// Stop any existing health check goroutine.
+		if origState.stopCh != nil {
+			close(origState.stopCh)
```

Comment:
We already lock the checker before doing this. There's no point to locking the state object since it is discarded within the checker lock (the map gets set to the new state object below).